File: ios/HybridDocumentScanner.swift
import Foundation
import NitroModules
import AVFoundation

class HybridDocumentScanner: HybridDocumentScannerSpec {
  func getNativeVersion() throws -> String {
    return "iOS AVFoundation & Vision Native Engine v1.0.0"
  }

  func ping(message: String) throws -> String {
    return "iOS Swift received: '\(message)'"
  }

  func getCameraPermissionStatus() throws -> String {
    switch AVCaptureDevice.authorizationStatus(for: .video) {
    case .authorized:
      return "granted"
    case .denied:
      return "denied"
    case .restricted:
      return "restricted"
    case .notDetermined:
      return "not-determined"
    @unknown default:
      return "not-determined"
    }
  }

  func requestCameraPermission() throws -> Promise<Bool> {
    let promise = Promise<Bool>()
    let status = AVCaptureDevice.authorizationStatus(for: .video)
    switch status {
    case .authorized:
      promise.resolve(withResult: true)
    case .denied, .restricted:
      promise.resolve(withResult: false)
    case .notDetermined:
      AVCaptureDevice.requestAccess(for: .video) { granted in
        promise.resolve(withResult: granted)
      }
    @unknown default:
      promise.resolve(withResult: false)
    }
    return promise
  }

  func capturePhoto(options: NativeCaptureOptions) throws -> Promise<NativeCapturedDocument> {
    let promise = Promise<NativeCapturedDocument>()

    DispatchQueue.main.async {
      guard let cameraView = DocumentCameraView.sharedCurrentView else {
        promise.reject(withError: NSError(
          domain: "HybridDocumentScanner",
          code: -1,
          userInfo: [NSLocalizedDescriptionKey: "Camera view active instance not found"]
        ))
        return
      }

      let enableFlash = options.enableFlash ?? false
      let autoCrop = options.autoCrop ?? false
      let detectPerspective = options.detectPerspective ?? false
      let documentType = options.documentType ?? "cccd"

      cameraView.capture(enableFlash: enableFlash, autoCrop: autoCrop, detectPerspective: detectPerspective, documentType: documentType) { result in
        switch result {
        case .success(let dict):
          let doc = NativeCapturedDocument(
            imageUri: dict["imageUri"] as? String ?? "",
            width: dict["width"] as? Double ?? 0.0,
            height: dict["height"] as? Double ?? 0.0,
            orientation: dict["orientation"] as? Double ?? 0.0,
            isCropped: dict["isCropped"] as? Bool ?? false,
            corners: nil
          )
          promise.resolve(withResult: doc)
        case .failure(let error):
          promise.reject(withError: error)
        }
      }
    }

    return promise
  }

  func validateDocumentImage(imageUri: String) throws -> Promise<ImageValidationResult> {
    let promise = Promise<ImageValidationResult>()
    DispatchQueue.global(qos: .userInitiated).async {
      let res = ImageValidator.validate(imageUri: imageUri)
      let result = ImageValidationResult(hasFace: res.hasFace, imageHash: res.hash)
      promise.resolve(withResult: result)
    }
    return promise
  }

  func compareImages(imageUri1: String, imageUri2: String) throws -> Promise<Double> {
    let promise = Promise<Double>()
    DispatchQueue.global(qos: .userInitiated).async {
      let sim = ImageValidator.compare(imageUri1: imageUri1, imageUri2: imageUri2)
      promise.resolve(withResult: sim)
    }
    return promise
  }

  func cleanCache() throws -> Promise<Bool> {
    let promise = Promise<Bool>()
    DispatchQueue.global(qos: .utility).async {
      let success = DocumentCameraView.clearCacheFiles()
      promise.resolve(withResult: success)
    }
    return promise
  }
}

File: android/src/main/java/com/margelo/nitro/rndocumentscanner/HybridDocumentScanner.kt
package com.margelo.nitro.rndocumentscanner

import android.Manifest
import android.content.pm.PackageManager
import androidx.annotation.Keep
import androidx.core.content.ContextCompat
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.rndocumentscanner.camera.DocumentCameraManagerView
import com.rndocumentscanner.utils.ImageValidator

@Keep
class HybridDocumentScanner : HybridDocumentScannerSpec() {
  override fun getNativeVersion(): String {
    return "0.1.0-android"
  }

  override fun ping(message: String): String {
    return "Android Pong: $message"
  }

  override fun getCameraPermissionStatus(): String {
    val context = NitroModules.applicationContext
    if (context != null) {
      val status = ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA)
      if (status == PackageManager.PERMISSION_GRANTED) {
        return "granted"
      }
    }
    return "denied"
  }

  override fun requestCameraPermission(): Promise<Boolean> {
    val context = NitroModules.applicationContext
    if (context != null) {
      val status = ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA)
      if (status == PackageManager.PERMISSION_GRANTED) {
        return Promise.resolved(true)
      }
    }
    return Promise.resolved(false)
  }

  override fun capturePhoto(options: NativeCaptureOptions): Promise<NativeCapturedDocument> {
    return Promise.async {
      val cameraView = DocumentCameraManagerView.sharedCurrentView
        ?: throw Exception("Camera view active instance not found on Android")

      val enableFlash = options.enableFlash ?: false
      val autoCrop = options.autoCrop ?: false
      val detectPerspective = options.detectPerspective ?: false
      val documentType = options.documentType ?: "cccd"

      kotlin.coroutines.suspendCoroutine { continuation ->
        cameraView.post {
          cameraView.capturePhoto(enableFlash, autoCrop, detectPerspective, documentType) { result ->
            result.fold(
              onSuccess = { map ->
                val doc = NativeCapturedDocument(
                  imageUri = map["imageUri"] as String,
                  width = map["width"] as Double,
                  height = map["height"] as Double,
                  orientation = map["orientation"] as Double,
                  isCropped = map["isCropped"] as Boolean,
                  corners = null
                )
                continuation.resumeWith(Result.success(doc))
              },
              onFailure = { error ->
                continuation.resumeWith(Result.failure(error))
              }
            )
          }
        }
      }
    }
  }

  override fun validateDocumentImage(imageUri: String): Promise<ImageValidationResult> {
    return Promise.async {
      val res = ImageValidator.validate(imageUri)
      ImageValidationResult(res.hasFace, res.hash)
    }
  }

  override fun compareImages(imageUri1: String, imageUri2: String): Promise<Double> {
    return Promise.async {
      ImageValidator.compare(imageUri1, imageUri2)
    }
  }
}
